doc/design/audit-trail-mvp.md

```diff
@@ -0,0 +1,152 @@
+
+## Requirements
+
+1. **Persistent tracking of significant actions.** Whenever a *significant
+   action* (see list [below](#significant-actions)) occurs successfully, the archive will generate a
+   persistent record of that action, including the following data:
+    - The date and time
+    - The authorized user who triggered the action
+    - The ID of the Dandiset on which the action occurs
+    - Event-specific data about the action itself
+
+2. **Display of audit log (to owners and admins only).** The Dandiset landing
+   page will display a rendered (i.e. human-readable) version of the log event
+   data (as long as the viewing user is an owner of the Dandiset, or an admin).
+
+   A possible example of such a display follows (the final rendered form may
+   differ from this depending on best practices, etc.):
+    
+    ```
+    YYYYMMDDTHHMMSS.mmmm: X added asset at path (/full/path) checksum
+    YYYYMMDDTHHMMSS.mmmm: X modified asset at path (/full/path) checksum
+    YYYYMMDDTHHMMSS.mmmm: X deleted asset at path (/full/path) checksum
+    YYYYMMDDTHHMMSS.mmmm: X modified dandiset metadata
+    ```
+    
+3. **Extensible significant action list.** The significant actions the system is
+   capable of tracking should be extensible--that is, in the future new actions
+   should be able to be added to the system, and existing actions should be able
+   to expand the specific data they carry.
+
+## Significant Actions
+
+The following is the list of actions performed on Dandisets that are considered
+"significant" (i.e., require an audit record to be generated), along with any
+additional, event-specific data each will carry with it.
+
+|  | General data | Specific data |
+| --- | --- | --- |
+| Create Dandiset | date_time, user_id, dandiset_id | metadata |
+| Add owner | date_time, user_id, dandiset_id | owner_id |
+| Remove owner | date_time, user_id, dandiset_id | owner_id |
+| Update Dandiset metadata | date_time, user_id, dandiset_id | metadata |
+| Add asset | date_time, user_id, dandiset_id | asset_path, checksum, metadata, blob_id |
+| Update asset | date_time, user_id, dandiset_id | asset_path, checksum, blob_id, metadata |
+| Remove asset | date_time, user_id, dandiset_id | asset_path |
+| Publish Dandiset | date_time, user_id, dandiset_id | version |
+| Delete Dandiset | date_time, user_id, dandiset_id | none |
+
+## Asset chain analysis
+
+This section provides some analysis of the historical data living in the
+"`Asset` chains" kept in the `Asset.previous` links. The immediate goal is to
+determine how much "significant action" data can be reconstructed therefrom;
+there is also some analysis of how this may affect provenance work in the
+future.
+
+For context: because most changes to an `Asset` model cause a new model instance
+to be created, and because that new `Asset` instance points to its "ancestor"
+via a `previous` property, the history of an asset is at least partially
+traceable through this "asset chain".
```

Comment:
I don't really understand this section. I suppose you are talking about previous versions of assets that belong to published dandisets to these previous versions are still kept around even as the asset is modified? Is that correct?

Reply:
What sorts of "analysis" are you thinking of?

Reply:
This is more of a technical quirk that presaged this feature (and provenance concerns).

Essentially, we're doing something kind of odd whenever an `Asset` is changed: we manufacture an entirely new `Asset` object, then refer to the original `Asset` object through the new `Asset`'s `previous` value. This was done to attempt to preserve the historical record, and this section analyzes how well that has worked, and what we can do to shed those old `Asset` objects via garbage collection.

This is different from hanging onto assets that belong to published versions--those will not be affected by this design.

Reply:
> What sorts of "analysis" are you thinking of?

The analysis is here in this design doc, not something that becomes a feature of the system (apologies if that wasn't clear). In particular, this analysis has led to the Operations plan detailed in the final section of the doc. The main upshot there is that if we take a snapshot of the database as it is at the moment we turn on the new audit feature, we can still, at any point in the future, reconstruct whatever history is recoverable from that snapshot--or we can just live with the fact that we don't have a complete history before that moment. It's up to us, since there's no loss of information and therefore no urgency.

Reply:
(@bendichter, I'm going to resolve this conversation; please unresolve and continue the conversation if you want to discuss further.)